### PR TITLE
[BIP44-33] AddressPool - core types and algorithm to support address …

### PR DESCRIPTION
**Pull request by uroboros**
…discovery for EOS/FOS wallets

# Linked Issue

https://github.com/input-output-hk/cardano-wallet/issues/33

# Comments

First draft of AddressPool module, sans any tests.

### CODE REVIEW

File: src/Cardano/Wallet/Kernel/HdSeq.hs [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+         selectOurs
+       ,
+         AccountPoolStatus (..)
+       ,
+         AccountPool (..)
+       ,
+         AddressPool (..)
+       ,
+         AccountPoolAmendment (..)
+       ) where
```
Comment (KtorZ) [inline src/Cardano/Wallet/Kernel/HdSeq.hs:24]: so
many 
new
lines

:smile:

File: src/Cardano/Wallet/Kernel/HdSeq.hs [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+
+{-------------------------------------------------------------------------------
+  AddressPool Types
+-------------------------------------------------------------------------------}
+
+-- | When an address is discovered in a sequence of addresses, if that
+--   address is "close enough" to the end of the sequence, we want to generate
+--   more addresses in the sequence (to increase the range of addresses that
+--   may be discovered). The AddressPoolGap expresses this notion of "close enough".
+type AddressPoolGap = Word32
```
Comment (KtorZ) [inline src/Cardano/Wallet/Kernel/HdSeq.hs:63]: _may fix_

`Word32` sounds like a lot. This has some overlap with #85; so I'd suggest to align with @denisshevchenko and either conflate the two modules into one, or have this one import the AddressPoolGap type.

File: src/Cardano/Wallet/Kernel/HdSeq.hs [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+       ,
+         AccountPoolAmendment (..)
+       ) where
+
+import           Universum
+
+import qualified Data.Map as Map
+import qualified Data.List.NonEmpty as NE
+
+import           Cardano.Wallet.Kernel.DB.HdWallet(HdAddressIx (..))
```
Comment (KtorZ) [inline src/Cardano/Wallet/Kernel/HdSeq.hs:31]: _must fix_

This one doesn't belong here I believe.

File: src/Cardano/Wallet/Kernel/HdSeq.hs [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+-------------------------------------------------------------------------------}
+
+-- | Construct a sequenced range of addresses, starting at index 'fromIx'
+-- TODO @uroboros PRECONDITION - range >= 2
+newAddressRange :: AddressPoolGap
+                -> HdAddressIx
+                -> NewAddress address
+                -> NonEmpty (address,HdAddressIx)
+newAddressRange range (HdAddressIx fromIx) newAddress
+    = (f fromIx) NE.:| (map f [fromIx + 1 .. fromIx + range - 1])
```
Comment (KtorZ) [inline src/Cardano/Wallet/Kernel/HdSeq.hs:143]: _may fix_ 

No need for qualified import here, `:|` is common enough to be used as a operator.

File: src/Cardano/Wallet/Kernel/HdSeq.hs [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+      ,
+        _addrPoolGap        = gap
+      ,
+        -- the constructor to extend the address pool range for an account
+        _addrPoolNewAddress = newAddress
+      ,
+        -- the constructor to create the next account in a sequence (given the
+        -- previous account)
+        _addrPoolNewAccount = Just newAccount
+    }
```
Comment (KtorZ) [inline src/Cardano/Wallet/Kernel/HdSeq.hs:253]: _must fix_

Let's not bother about FOS wallet now and proceed in steps. For now, we only consider case where we ignore relationship accounts have to each others and we only build from a set of account public keys.

File: src/Cardano/Wallet/Kernel/HdSeq.hs [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+--
+--   In other words, the address pools we match against may change as we
+--   discover new addresses.
+addrsInPools :: (Eq account, Ord address)
+             => AddressPoolGap
+             -> Maybe (NewAccount account)
+             -> NewAddress address
+             -> [AccountPool account address]
+             -> [address]
+             -> [(account, HdAddressIx)]
```
Comment (KtorZ) [inline src/Cardano/Wallet/Kernel/HdSeq.hs:358]: _must fix_

This looks a lot complicated for an exposed function. One of the goal we want to achieve by making an isolated module is to provide some easy handles on an opaque type such that, it's easy for users of this module to achieve what they want. 

```hs
addrInPool 
  :: AddressPool account address
  -> address
  -> (AddressPool account address, Maybe (account, AddressIdx))
addrInPool ... =
```

That's all we should need, right ?

File: src/Cardano/Wallet/Kernel/HdSeq.hs [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+       ,
+         selectOurs
+       ,
+         AccountPoolStatus (..)
+       ,
+         AccountPool (..)
+       ,
+         AddressPool (..)
+       ,
+         AccountPoolAmendment (..)
```
Comment (KtorZ) [inline src/Cardano/Wallet/Kernel/HdSeq.hs:23]: _must fix_

We expose too many things from this module. The point of making this an isolated module is to encapsulate the complexity and implementation details inside the module and make sure they don't get out! 
Underneath it all, the API boils down to this:

```hs
module 
  ( AddressPool
  , initAddressPool
  , addrInPool -- cf comment below
  ) where
```

We may provide some extra "debugging" / testing helpers like "totalPoolSize" to help us write properties about those. But, conceptually, I believe we don't need to expose more than that, don't we?

File: src/Cardano/Wallet/Kernel/HdSeq.hs [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+
+-- | Constructor for creating a new sequential Hd address, takes as argument
+--   the address index
+type NewAddress address = HdAddressIx -> address
+
+-- | The pool of "our" addresses for an account
+
+-- Note: the account and address types are polymorphic in order to support
+-- varying wallet types
+data AccountPool account address = AccountPool {
```
Comment (KtorZ) [inline src/Cardano/Wallet/Kernel/HdSeq.hs:82]: _may fix_ 

What above shoving in there the `Ord` and `Eq` instances for `account` and `address` ?

**Review by KtorZ — CHANGES_REQUESTED** [review]
- Let's not even mention FOS wallets at this stage
- We ought to provide a "dead simple" API for this module and have the complexity lives inside it

Haven't much reviewed the algorithm so-to-speak as I think the various comment will provoke some important refactoring :/ 
Also, make sure to align with @denisshevchenko on the `AddressPoolGap` type.

**Comment by denisshevchenko** [comment]
Closed, because simplified version is here: https://github.com/input-output-hk/cardano-wallet/pull/118

**Pull request closed without merging**